src/apps/api/views/competitions.py

```diff
@@ -347,9 +368,27 @@ def perform_create(self, serializer):
         self._ensure_organizer_participants_accepted(instance)
 
     def perform_update(self, serializer):
+        instance = self.get_object()
+
+        initial_tasks = {}
+        for phase in instance.phases.all():
+            initial_tasks.update({phase.id: set(phase.tasks.all())})
+
         instance = serializer.save()
         self._ensure_organizer_participants_accepted(instance)
 
+        saved_tasks = {}
+        new_phases = {}
+        for phase in instance.phases.all():
+            saved_tasks.update({phase.id: set(phase.tasks.all())})
+            new_phases.update({phase.id: phase})
+        for key in saved_tasks.keys():
+            if key in initial_tasks.keys():
+                print(saved_tasks[key] - initial_tasks[key])
+                new_tasks = list(saved_tasks[key] - initial_tasks[key])
+                if new_tasks:
+                    self.run_new_task_submissions(new_phases[key], new_tasks)
```

Comment:
I don't think we need this `new_phases` dictionary, since we could get the phase off of any task in the phase. This is another opportunity to do another dictionary comprehension & `prefetch_related` as my comment above. The call to `filter(pk__in=initial_tasks.keys())` removes the requirement for the extra check if the phase is in `initial_tasks`. There are a lot of changes here, so let me know if you have any questions. These are mainly just code style improvements. I believe that the logic of your solution is correct. Note the following suggestion:
```suggestion
        saved_tasks = {phase.id: set(phase.tasks.filter(pk__in=initial_tasks.keys())) \
        for phase in instance.phases.all().prefetch_related('tasks')}
        for phase_id in saved_tasks:
            new_tasks = list(saved_tasks[phase_id] - initial_tasks[phase_id])
            if new_tasks:
                self.run_new_task_submissions(new_tasks[0].phase, new_tasks)
```

Reply:
Final Iteration (Thanks for the help! Much cleaner)

```python
        saved_tasks = {phase.id: set(phase.tasks.all()) for phase in instance.phases.filter(pk__in=initial_tasks.keys()).prefetch_related('tasks')}
        for phase_id in saved_tasks:
            new_tasks = list(saved_tasks[phase_id] - initial_tasks[phase_id])
            if new_tasks:
                self.run_new_task_submissions(instance.phases.get(pk=phase_id), new_tasks)

```

Reply:
Looks good!